spring-cloud-gcp-pubsub/src/test/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubInboundChannelAdapterTests.java

```diff
@@ -16,28 +16,171 @@
+		this.countDownLatch = new CountDownLatch(1);
+
+		NackAnswer nackAnwser = new NackAnswer(this.returnValue);
+
+		this.ackReplyConsumer = mock(AckReplyConsumer.class);
+
+		doAnswer(nackAnwser).when(this.ackReplyConsumer).nack();
+
+		this.testAnswer = new TestAnswer(this.countDownLatch, this.ackReplyConsumer);
+	}
 
 	@Test(expected = IllegalArgumentException.class)
 	public void testNonNullAckMode() {
 		PubSubInboundChannelAdapter adapter = new PubSubInboundChannelAdapter(
-				this.pubSubTemplate, "testSubscription");
+				this.pubSubOperations, "testSubscription");
+
 		adapter.setAckMode(null);
 	}
+
+	@Test
+	public void testAckModeAuto() throws Exception {
+		PubSubInboundChannelAdapter adapter = new PubSubInboundChannelAdapter(
+				this.pubSubSubscriberOperations, "testSubscription");
+
+		adapter.setAckMode(AckMode.AUTO);
+
+		when(this.pubSubSubscriberOperations.subscribe(anyString(), any(MessageReceiver.class))).then(this.testAnswer);
+
+		adapter.setOutputChannel(this.messageChannel);
+
+		try {
+			adapter.start();
+		}
+		catch (Throwable t) {
+
+		}
+
+		Assert.assertTrue(this.countDownLatch.await(10, TimeUnit.SECONDS));
+		Assert.assertEquals(NACK, this.returnValue.get());
+	}
+
+	@Test
+	public void testAckModeAutoAck() throws Exception {
+		PubSubInboundChannelAdapter adapter = new PubSubInboundChannelAdapter(
+				this.pubSubSubscriberOperations, "testSubscription");
+
+		adapter.setAckMode(AckMode.AUTO_ACK);
+
+		when(this.pubSubSubscriberOperations.subscribe(anyString(), any(MessageReceiver.class))).then(this.testAnswer);
+
+		adapter.setOutputChannel(this.messageChannel);
+
+		try {
+			adapter.start();
+		}
+		catch (Throwable t) {
+
+		}
+
+		Assert.assertTrue(this.countDownLatch.await(10, TimeUnit.SECONDS));
+		Assert.assertNull(this.returnValue.get());
+	}
+
+	private class TestAnswer implements Answer<MessageReceiver> {
+
+		private CountDownLatch countDownLatch;
+
+		private AckReplyConsumer ackReplyConsumer;
+
+		TestAnswer(CountDownLatch countDownLatch, AckReplyConsumer ackReplyConsumer) {
+			this.countDownLatch = countDownLatch;
+			this.ackReplyConsumer = ackReplyConsumer;
+		}
+
+		public MessageReceiver answer(InvocationOnMock invocation) throws Throwable {
+			try {
+				PubsubMessage pubsubMessage = PubsubMessage.newBuilder().setData(
+						ByteString.copyFrom("Testing 1 2 3".getBytes("UTF-8"))).build();
+
+				MessageReceiver messageReceiver = invocation.getArgument(1);
+				messageReceiver.receiveMessage(pubsubMessage, this.ackReplyConsumer);
+			}
+			finally {
+				this.countDownLatch.countDown();
+			}
+
+			return null;
+		}
+	}
+
+	private class NackAnswer implements Answer<Boolean> {
+
+		private AtomicReference<String> returnValue;
+
+		NackAnswer(AtomicReference<String> returnValue) {
+			this.returnValue = returnValue;
+		}
+
+		public Boolean answer(InvocationOnMock invocationOnMock) throws Throwable {
+			this.returnValue.set(NACK);
+
+			return null;
```

Comment:
I don't see reason in the separate class for such a simple logic in body. More over it is much cleaner now with Java 8 lambdas.